### PR TITLE
VM: improve supports for using marketplace images

### LINKED ISSUE #3411: Support market-place ordering
### Description
**Outline the issue here:**
When I try to install VM from Azure Marketplace via Azure CLI 2.0 I get the error message:
At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details. {
  "error": {
    "code": "ResourcePurchaseValidationFailed",
    "message": "User failed validation to purchase resources. Error message: 'Legal terms have not been accepted for this item on this subscription. To accept legal terms, please go to the Azure portal (http://go.microsoft.com/fwlink/?LinkId=534873) and configure programmatic deployment for the Marketplace item or create it there for the first time'"
  }
}  Correlation ID: 5d540903-63f7-4d79-8626-92ae748f6d2e
In order to overcome it I have to login to portal.azure.com and configure programmatic deployment there. After that I can create VM with no problem.
I wonder if it's possible to configure programmatic deployment via Azure CLI in order not switch from CLI to portal and back again?
thanks!

---

### Environment summary

**Install Method:** How did you install the CLI? (e.g. pip, interactive script, apt-get, Docker, MSI, nightly)  
Answer here:
apt-get
**CLI Version:** What version of the CLI and modules are installed? (Use `az --version`)  
Answer here:
root@ALEXEY-HP:~# az --version
azure-cli (2.0.6)

acr (2.0.4)
acs (2.0.6)
appservice (0.1.6)
batch (2.0.4)
cdn (0.0.2)
cloud (2.0.2)
cognitiveservices (0.1.2)
command-modules-nspkg (2.0.0)
component (2.0.4)
configure (2.0.6)
core (2.0.6)
cosmosdb (0.1.6)
dla (0.0.6)
dls (0.0.6)
feedback (2.0.2)
find (0.2.2)
interactive (0.3.2)
iot (0.1.5)
keyvault (2.0.4)
lab (0.0.4)
monitor (0.0.4)
network (2.0.6)
nspkg (3.0.0)
profile (2.0.4)
rdbms (0.0.1)
redis (0.2.3)
resource (2.0.6)
role (2.0.4)
sf (1.0.1)
sql (2.0.3)
storage (2.0.6)
vm (2.0.6)

Python (Linux) 3.6.1 (default, May 18 2017, 14:59:59)
[GCC 4.8.4]

Python location '/opt/az/bin/python3'
**OS Version:** What OS and version are you using?  
Answer here:
D:\>ver
Microsoft Windows [Version 10.0.14393]
64-bit

**Shell Type:** What shell are you using? (e.g. bash, cmd.exe, Bash on Windows)  
Answer here:
cmd.exe, Bash on Windows

### PR DESCRIPTION
**Pull request by yugangw-msft**
2 things:
1. Through the new command of `az vm image accept-term`, users can accept the term to use some market-place images with its own billing. Before, they have to do it through portal. This fixes #3411.
2. Update `az vm image show` to support `--urn`, so users can supply the urn returned from `az vm image list` and get the billing plan information easily

The flow looks like
```
urn = $(az vm image list --all --publisher kemptech --offer vlm-azure --sku basic-byol --query '[0].urn')
az vm image accept-term --urn $urn
az vm create -g rg -n vm1 --image $urn
```
### General Guidelines

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/5378
This is an **experimental preview** for @microsoft users.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/commands.py [hunk 1]
```diff
@@ -227,11 +227,12 @@ def load_command_table(self, _):
         g.custom_command('list', 'list_vm_extension_images')
 
     with self.command_group('vm image', compute_vm_image_sdk) as g:
-        g.command('show', 'get', exception_handler=empty_on_404)
         g.command('list-offers', 'list_offers')
         g.command('list-publishers', 'list_publishers')
         g.command('list-skus', 'list_skus')
         g.custom_command('list', 'list_vm_images')
+        g.custom_command('accept-term', 'accept_market_ordering_term')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/commands.py]: I think the command name here should be "accept-terms".

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py [hunk 2]
```diff
@@ -1217,6 +1218,46 @@ def list_vm_images(cmd, image_location=None, publisher_name=None, offer=None, sk
...
+    from azure.mgmt.marketplaceordering import MarketplaceOrderingAgreements
+
+    usage_err = 'usage error: --plan STRING --offer STRING --publish STRING |--urn STRING'
+    if urn:
+        if any([publisher, offer, plan]):
+            raise CLIError(usage_err)
+        publisher, offer, _, _ = urn.split(':')
+        image = show_vm_image(cmd, urn)
+        if not image.plan:
+            logger.warning('No plan information is associated with the image, hence not term to accept')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py]: Recommend: "Image {blah} has no terms to accept."

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 3]
```diff
@@ -344,6 +344,22 @@ def test_custom_image_with_plan(self, resource_group):
...
+    @ResourceGroupPreparer()
+    def test_vm_create_with_market_place_image(self, resource_group, resource_group_location):
+        self.kwargs.update({
+            'location': resource_group_location,
+            'publisher': 'kemptech',
+            'offer': 'vlm-azure',
+            'sku': 'basic-byol',
+            'plan': 'basic-byol'
+        })
+        urn = '{}:{}:{}:7.2.362142710'.format(self.kwargs['publisher'], self.kwargs['offer'], self.kwargs['sku'])
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: Recommend: `self.kwargs['urn'] = '{publisher}:{offer}:{sku}:7.12.blahblah'.format(**self.kwargs)`

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 4]
```diff
@@ -344,6 +344,22 @@ def test_custom_image_with_plan(self, resource_group):
...
+        self.cmd('vm image show --urn ' + urn, checks=self.check('plan.name', self.kwargs['plan']))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: Recommend: `vm image show --urn {urn}, checks=self.check('plan.name', '{plan}')`

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 5]
```diff
@@ -344,6 +344,22 @@ def test_custom_image_with_plan(self, resource_group):
...
+        self.cmd('vm image accept-term -p {publisher} --offer {offer} --plan {plan}', checks=self.check('accepted', True))
+        # repeat the same command using --urn
+        self.cmd('vm image accept-term --urn ' + urn, checks=self.check('accepted', True))
+        self.cmd('vm create -g {rg} -n vm1 --no-wait --image ' + urn)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: Same.  Recommend putting urn into kwargs and using the normal syntax to make it a little cleaner.

File: src/command_modules/azure-cli-vm/HISTORY.rst [hunk 6]
```diff
@@ -5,6 +5,7 @@ Release History
 
 2.0.25
 ++++++
+* vm image: support accept market terms to use vm images
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/HISTORY.rst]: Recommend: Added `vm image accept-terms` command to support using certain marketplace images.

**Comment by bernadinm** [comment]
Do you know by any chance when this will land on the master branch rather than then dev? Great work @yugangw-msft, this is a very helpful feature indeed!

**Comment by bernadinm** [comment]
Actually, the main reason why I cannot see it is that when doing `brew upgrade az`, the version it provided to me is this:

```
$ az -v
\azure-cli (2.0.12)

acr (2.0.9)
acs (2.0.11)
appservice (0.1.11)
batch (3.0.3)
billing (0.1.3)
cdn (0.0.6)
cloud (2.0.7)
cognitiveservices (0.1.6)
command-modules-nspkg (2.0.1)
component (2.0.6)
configure (2.0.10)
consumption (0.1.3)
container (0.1.7)
core (2.0.12)
cosmosdb (0.1.11)
dla (0.0.10)
dls (0.0.11)
feedback (2.0.6)
find (0.2.6)
interactive (0.3.7)
iot (0.1.10)
keyvault (2.0.8)
lab (0.0.9)
monitor (0.0.8)
network (2.0.11)
nspkg (3.0.1)
profile (2.0.9)
rdbms (0.0.5)
redis (0.2.7)
resource (2.0.11)
role (2.0.9)
sf (1.0.5)
sql (2.0.8)
storage (2.0.11)
vm (2.0.11)

Python (Darwin) 2.7.10 (default, Jul 15 2017, 17:16:57)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]

Python location '/Users/mbernadin/lib/azure-cli/bin/python'
```

I would imagine that this commit resides in a more recent version. I'll go pull it from docker directly. 

Thanks!

**Pull request merged**